### ISSUE TITLE
The Server should separate the counselor entries into objects before saving to the database.

### ISSUE THREAD

**Issue by Redmega** [Redmega/licensed-mental-health-counselors-browser#9]
Continuing from #8, we should split the counselor entries into objects with keys before saving to the lowdb. This will make instantiating the Practitioner class on the client side much cleaner, and will pave the way for category based searching and contact card UI (#5 and #1)

**Comment by Redmega** [comment]
@dsych Thanks for your contributions to the previous issue! Here's the continuation of it, if you want to work on it as well. I've already marked it as [In Progress] for you.

**Comment by dsych** [comment]
Here is what I am thinking: we certainly don't want to maintain two separate objects that are describing the `Practitioner`, so we can have a base class that describes properties needed for the json. The base will be an es6 class (stored in `.js` file inside `/server/lib`) and will receive an array which it will proceed to parse or nothing. The derived class will be a typescript class that is stored inside `src/components/practitioner.ts`. It will extend the base class and attache additional properties like `address`, `phone` etc. 
This way, once we change anything inside the base class, the db, json signatures and the client side signatures are all in sync.
It will look something like this:
```js
// /server/lib/practitioner.js
class Practitioner {
    constructor(raw) {
        if (raw && raw.length) {

            this.licId = raw[0];
            this.expDate = raw[1];
            this.orgDate = raw[2];
            this.licenseNum = raw[3];
            this.statusDate = raw[4];
            this.boardAction = raw[5];
            ...
            this.prescriptionIndicator = raw[30];
            this.dispensingIndicator = raw[31];
        } else {
            this.licId = "";
            this.expDate = "";
            this.orgDate = "";
            this.licenseNum = "";
            this.statusDate = "";
            this.boardAction = "";
            ...
            this.prescriptionIndicator = "";
            this.dispensingIndicator = "";
        }
    }
}
```

```js
import { Practitioner as P } from "../../server/lib/practitioner"

// /src/components/practitioner.ts
export class Practitioner extends P {
    [x: string]: any;
 
    phone: Phone;

    mailingAddress: Address;
    practiceAddress: Address;

    constructor(src) {
        super();
        Object.keys(src).forEach(key => {
            this[key] = src[key];
        });
        this.mailingAddress = {
            line1: this.mailingAddressLine1
            ...
        };
        this.phone = {
            phoneNum: this.phoneNumber
            ...
        }
    }

}
```

**Comment by Redmega** [comment]
That looks solid to me.

Another option might be to just use typescript on the server!

For development we can use `ts-node` as the executable for `nodemon`, and for prod we can just compile with babel / typescript into `build/` and use `"start": "node build/server.js"`

**Comment by dsych** [comment]
If migrating to TS is an option, I think we should do that instead. Makes sense to have server and client code living side by side on the same stack.
Once that is done, I can do proper inheritance between objects as per TS standards. Even though what I proposed above works, I hate the use of call type signatures in this way.

**Comment by Redmega** [comment]
Yeah, definitely an option. I can create a separate ticket for that and work on it tonight, or if you'd like you can take that on as well!

**Comment by dsych** [comment]
You can go ahead.

**Comment by Redmega** [comment]
Ended up getting home late yesterday. Will work on it tonight!

**Comment by Redmega** [comment]
OK, #10 is closed. Had some issues extending the Express Request -- if you're able to get the _declaration merging_ working properly, I'd be super appreciative! In the meantime we just had to specifically type the req param in our middlewares that access `.db`.

**Comment by Redmega** [comment]
From my perspective, looks like it's working properly (at least to vs-code), it just isn't being picked up by the typescript compiler: 
<img width="543" alt="screen shot 2018-10-24 at 10 10 52 pm" src="https://user-images.githubusercontent.com/662951/47471736-bb985d80-d7d9-11e8-9766-099842fbf226.png">

**Comment by Redmega** [comment]
So, I did _something_...

https://github.com/Redmega/licensed-mental-health-counselors-browser/commit/4187da4fa88da1adaa89f7ce7ad77cb041daf732

I don't think that's correct, but I'm fine with it for now since it's working.

**Comment by dsych** [comment]
Np, @Redmega, I will look into it tomorrow.

**Comment by Redmega** [comment]
So saw this entry, definitely confirms that we need to show both line 1 and line 2.

<img width="356" alt="screen shot 2018-10-25 at 1 14 43 am" src="https://user-images.githubusercontent.com/662951/47477473-68cb9f80-d7f3-11e8-996b-d35b6171116e.png">

**Comment by Redmega** [comment]
:wave: Made any progress on this yet?

I know hacktoberfest is over so if you're not so inclined I can take over this issue :)
